content/blog/2023-12-11-rvv-llvm-gisel/index.md

```diff
@@ -0,0 +1,177 @@
+```
+<vscale x VF x primitive data type>
+```
+where `VF` is the vectorization/grouping factor, it specifies how many vector elements are there in this group; a primitive data type; and `vscale` is a constant multiple of the specified number of vector elements, and `vscale` is unknown at compile time. For example, `<vscale x 4 x i32>` represents a scalable vector, which is a multiple of 4 32-bit integer values.
+
+
+Since there are four passes in the LLVM GlobalISel core pipeline, my design and implmentation are also based on these four passes.
+
+## IRTranslator
+IRTranslator is the first pass of GlobalISel pipeline. The goal of this pass is to translate LLVM IR into gMIR. And the essential thing that is happening in this pass is call lowering according to the ABI rules of the target, which is RISC-V in our case. The proposal of vector calling convention can be found [here](https://github.com/riscv-non-isa/riscv-elf-psabi-doc/pull/389).
+
+If we recall our example, what we want to get out of the first pass is something like:
+```
+%vc:_(<vscale x 1 x i8>) = G_ADD %va:_(<vscale x 1 x i8>), %vb:_(<vscale x 1 x i8>)
+```
+
+To achieve this, we need to first support parsing scalable vectors and to implement three main parts in call lowering, namely `lowerReturn`, `lowerFormalArguments`, and `lowerCall`. The functionalities of these three main parts are as what their names have suggested:
+1. `lowerReturn` enables functions to return scalable vector types. To be specific, scalable vector types in the input LLVM IR programs can be lowered into physical vector registers MIR based on the RISC-V vector calling conventions for return instructions. To implement this, we just write a function that translate to a return MIR if the return value is a scalable vector. It used to fall back to SelectionDAG before we implement this, now we can directly enable it in GlobalISel. [This patch](https://github.com/llvm/llvm-project/pull/71587) implements this functionality.
+2. `lowerFormalArguments` supports passing scalable vectors as arguments. We do it by implementing a function that takes in type information and returns true/false based on the type and other relevant information, such as if the target hardware supports vector instructions, and so forth. The existing argument-lowering function can take this as a helper function and enables passing scalable vectors as function arguments. [This patch](https://github.com/llvm/llvm-project/pull/70882) supports lowering formal arguments.
+3. `lowerCall` is used to lower the given call instruction according to the ABI. 
```

Comment:
This is not yet implemented for vectors on RISCV since it was not a blocker to do CodeGen for vector add. Maybe you clarify?